### ISSUE TITLE
Sharing Code Between POS & Fannie

### ISSUE THREAD

**Issue by gohanman** [CORE-POS/IS4C#470]
In #466 @flathat points out that POS' BasicModel class cannot adjust a VARCHAR column's width where as Fannie's BasicModel class can. I want to spin this off as a separate issue because there's really a larger problem of mostly duplicate code in Fannie and POS with slight differences.

The "mostly duplicate" part likely results in some wasted time, and the "slight differences" part leads to unnecessary confusion for developers. Two classes with the same name and different behavior is going to lead to unexpected surprises. I'd like to start bridging this gap.

Currently, POS doesn't reference any code from the "fannie" directory tree and Fannie doesn't reference any code from the "pos/is4c-nf" directory tree. I think adding an additional top-level directory specifically for shared code is going to be better than creating a bunch of cross linkages.

From a practical standpoint this would probably means the end of fannieDevelop and posDevelop as separate branches to avoid conflicts in shared code. I've been thinking my branching strategy is wrong lately anyway. What I should be doing is creating a **new** development branch after each release rather than dragging along perpetual parallel develop branches. So for example when 1.2 is released, I'd create a new branch "dev-1.3" based on master and close the previous dev branch(es). I think the commit history would be a lot cleaner that way.

So, in summary, I'm proposing:
* New top level directory for common code used by both POS and Fannie
* New branching strategy with a single development branch for both POS and Fannie and a fresh branch following each release

**Comment by jonkeim** [comment]
A [root]/lib sort of directory is a very good idea.

A single dev branch also sounds like a good idea.  I can see how that would simplify the commit graph.

**Comment by jdpurdyvi** [comment]
A root lib directory should help with some of the overlap, and provide a place for endeavors that span the two sides of the project. Or even expand the project in new directions. Sounds good. 
dev-1.3-POS and dev-1.3-Fannie versus a combined dev-1.3 are still interesting branching topics. 
 


     On Wednesday, March 18, 2015 4:53 PM, Jon Keim <notifications@github.com> wrote:
   
 

 A [root]/lib sort of directory is a very good idea.A single dev branch also sounds like a good idea. I can see how that would simplify the commit graph.—

**Comment by gohanman** [comment]
Does the root lib then become a third dev branch? I expect commits to the root lib would get merged into both dev topic branches (or cherry-picked across them) more or less immediately. And then when the two dev topic branches are eventually merged into master the history will contain duplicate entries for the root lib commits.

**Comment by gohanman** [comment]
A single SQLManager class shared across POS and Fannie:
1874bc2e25d9ac5b34e878660dd95c5347673947

**Comment by joelbrock** [comment]
dang!  now we are getting somewhere!

**Comment by gohanman** [comment]
#472 is basically why I'd like to see a unified develop branch. A big PR like that really doesn't belong in master. It ought to be merged into some kind of development branch for testing and validation, but there's no upstream branch that provides a valid target. A PR effecting both POS and Fannie is stuck targeting master.

**Comment by gohanman** [comment]
After looking at some other repositories/projects, I think my branching idea was backwards. Development should be happening directly on master (with or without temporary feature branches). Each *release* should create its own branch. Back porting fixes into previous releases if necessary is simpler since you can go back to the release branch and add things without interfering with ongoing development on master.

**Comment by gohanman** [comment]
#479 reinforces the idea that we are (or I am) branching incorrectly. There's a bug in SQLManager. If I fix it in the current develop branch(es), anyone tracking master has to live with the bug until the next release. If I fix it in the develop branch(es) and master, there's going to be a merge conflict at the next release. In either case, someone tracking just releases is stuck with the bug until one of the current branches is stabilized, tested, and marked as a release.

If there was a branch named "version-1.2" and development was happening in master (with or without topic branches), I could fix the bug in both version-1.2 and ongoing development without creating a conflict. The version-1.2 branch is never getting merged back into ongoing development. The branch is just there to apply bugfixes on a stable(ish) release. More importantly, bugfixes could be released more quickly without waiting for a full feature development cycle. In theory this could even go back several versions so someone still following a version-1.1 or version-1.0 branch gets the fix too. For practical purposes there'd have to be some limit to the number of old releases receiving maintenance, but I think that would be really appealing from an end user standpoint. There's more leeway to upgrade at your own pace.

**Comment by gohanman** [comment]
Example: a version-1.2 branch with a bugfix following the release commit:
https://github.com/CORE-POS/IS4C/commits/version-1.2
